apps/CA.pl.in

```diff
@@ -127,59 +163,289 @@ if ($WHAT eq '-newcert' ) {
+        print STDERR "Signing of certificate failed\n";
+        exit 1;
+    }
+    # create a PKCS#12 file from the signed cert and key
+    if (-f "${CERTTOP}/private/${EXPORTPASS_FILE}") {
+        $PASSOPT = "-passout file:${CERTTOP}/private/${EXPORTPASS_FILE}";
+    } else {
+        print "Hint: putting PKCS#12 export password file '$EXPORTPASS_FILE' in\n"
+            . " ${CERTTOP}/private will ease management.\n";
+    }
+    pkcs12($cnametxt);
+    if ($RET != 0) {
+        print STDERR "Creation of PKCS#12 failed\n";
+        exit 1;
+    }
+    # tidy up the keys
+    move($NEWP12,"${CERTTOP}/private/${cname}.p12");
+    move($NEWKEY,"${CERTTOP}/private/${cname}.key");
+    move($NEWCERT,"${CERTTOP}/certs/${cname}.pem");
+    unlink $NEWREQ;
+    my $msg = <<"EOM";
+Keys generated successfully:
+${CERTTOP}/certs/${cname}.pem     public key(signed)
+${CERTTOP}/private/${cname}.key   private key
+${CERTTOP}/private/${cname}.p12   PKCS#12(the bag of above two)
+EOM
+    print $msg. "\n";
+    print "CA public keys are: ${CATOP}/${CACERT}(PEM format), ${CACERTDER}(DER)\n";
+}
+
+# Update Certificate Revocation List
+sub gencrl
+{
+    $RET = run("$CA $OPENSSL_CONFIG -gencrl -out ${CATOP}/crl/$CACRL");
+    if ($RET != 0) {
+        print STDERR "Failed to Generated CRL\n";
+        exit 1;
+    }
+    print "Generated CRL is in ${CATOP}/crl/$CACRL\n";
+    # Place what to do after CRL update, e.g.
+    # $RET = run("cd ${CERTTOP} && make");
+}
+
+# Revoke a certificate
+sub revoke
+{
+    # revoke a certificate
+    my ($cname,$reason) = @_;
+    unless ($cname) {
         print "Certificate filename is required; reason optional.\n";
         exit 1;
     }
-    my $reason = $ARGV[2];
-    $reason = " -crl_reason $reason"
-        if defined $reason && crl_reason_ok($reason);
-    $RET = run("$CA -revoke \"$cname\"" . $reason);
+    $reason = "" unless defined $reason;
+    if (length($reason) gt 0 && crl_reason_ok($reason)) {
+        $reason = " -crl_reason $reason"
+    }
+    unless (-f "$cname") {
+        print STDERR "No such file $cname\n";
+        exit 1;
+    }
+    $RET = run("$CA $OPENSSL_CONFIG -revoke \"$cname\"" . $reason);
+    if ($RET != 0) {
+        print STDERR "Failed to revoke $cname\n";
+        exit 1;
+    }
+}
+
+# Revoke and delete a set of cert, key and PKCS#12 files
+sub p12mng_revoke
+{
+    $OPENSSL_CONFIG = $OPENSSL_CONFIG_P12MNG unless $OPENSSL_CONFIG;
+    my ($cname,$reason) = @_;
+    unless ($cname) {
+        print "Certificate name is required; reason optional.\n";
+        exit 1;
+    }
+    $reason = "" unless defined $reason;
+    unless (-f "${CERTTOP}/certs/${cname}.pem") {
+        print STDERR "No such file ${cname}.pem\n";
+        exit 1;
+    }
+    revoke("${CERTTOP}/certs/${cname}.pem",$reason);
+    # remove the revoked key set
+    run("rm -i ${CERTTOP}/certs/${cname}.pem");
+    run("rm -i ${CERTTOP}/private/${cname}.key");
+    run("rm -i ${CERTTOP}/private/${cname}.p12");
```

Comment:
These removals aren't a good idea, generally speaking.  Revoked certificates can still be used to validate signatures, if that's what they were used for.